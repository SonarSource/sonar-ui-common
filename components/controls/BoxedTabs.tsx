/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { styled, ThemedProps } from '../theme';

export interface BoxedTabsProps<K> {
  onSelect: (key: K) => void;
  selected?: K;
  tabs: Array<{ key: K; label: React.ReactNode }>;
}

const baseBorder = ({ theme }: ThemedProps) => `1px solid ${theme.colors.barBorderColor}`;

const TabContainer = styled.div`
  display: flex;
  flex-direction: row;
`;

const StyledTab = styled.button<{ active: boolean }>`
  position: relative;
  background-color: ${props => (props.active ? 'white' : props.theme.colors.barBackgroundColor)};
  border-top: ${baseBorder};
  border-left: ${baseBorder};
  border-right: none;
  border-bottom: none;
  margin-bottom: -1px;
  min-width: 128px;
  min-height: 56px;
  cursor: pointer;
  outline: 0;
  padding: calc(2 * ${props => props.theme.sizes.gridSize});

  &:hover {
    background-color: ${props => props.theme.colors.barBackgroundColorHighlight};
  }

  &:last-child {
    border-right: ${baseBorder};
  }
`;

const ActiveBorder = styled.div<{ active: boolean }>`
  display: ${props => (props.active ? 'block' : 'none')};
  background-color: ${props => props.theme.colors.blue};
  height: 3px;
  width: 100%;
  position: absolute;
  left: 0;
  top: -1px;
`;

export default function BoxedTabs<K>(props: BoxedTabsProps<K>) {
  const { tabs, selected } = props;
  return (
    <TabContainer>
      {tabs.map(({ key, label }, i) => (
        <StyledTab
          active={selected === key}
          key={i}
          onClick={() => props.onSelect(key)}
          type="button">
          <ActiveBorder active={selected === key} />
          {label}
        </StyledTab>
      ))}
    </TabContainer>
  );
}
