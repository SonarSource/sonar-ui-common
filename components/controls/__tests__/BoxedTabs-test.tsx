/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import BoxedTabs, { BoxedTabsProps } from '../BoxedTabs';

it('should render correctly', () => {
  expect(mountRender()).toMatchSnapshot();
});

it('should call onSelect when a tab is clicked', () => {
  const onSelect = jest.fn();
  const wrapper = shallowRender({ onSelect });

  wrapper
    .find('Styled(button)')
    .get(1)
    .props.onClick();

  expect(onSelect).toHaveBeenCalledWith('b');
});

function shallowRender(overrides: Partial<BoxedTabsProps<string>> = {}) {
  return shallow(dom(overrides));
}

function mountRender(overrides: Partial<BoxedTabsProps<string>> = {}) {
  return mount(dom(overrides));
}

function dom(overrides) {
  return (
    <BoxedTabs
      onSelect={jest.fn()}
      selected="a"
      tabs={[
        { key: 'a', label: 'labela' },
        { key: 'b', label: 'labelb' },
        {
          key: 'c',
          label: (
            <span>
              Complex label <strong>!!!</strong>
            </span>
          )
        }
      ]}
      {...overrides}
    />
  );
}
