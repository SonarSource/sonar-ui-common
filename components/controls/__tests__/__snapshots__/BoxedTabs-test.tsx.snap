// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-1 {
  position: relative;
  background-color: white;
  border-top: 1px solid #e6e6e6;
  border-left: 1px solid #e6e6e6;
  border-right: none;
  border-bottom: none;
  margin-bottom: -1px;
  width: 128px;
  height: 56px;
  cursor: pointer;
  outline: 0;
}

.emotion-1:last-child {
  border-right: 1px solid #e6e6e6;
}

.emotion-0 {
  display: block;
  background-color: #4b9fd5;
  height: 3px;
  width: 100%;
  position: absolute;
  top: -1px;
}

.emotion-3 {
  position: relative;
  background-color: #f3f3f3;
  border-top: 1px solid #e6e6e6;
  border-left: 1px solid #e6e6e6;
  border-right: none;
  border-bottom: none;
  margin-bottom: -1px;
  width: 128px;
  height: 56px;
  cursor: pointer;
  outline: 0;
}

.emotion-3:last-child {
  border-right: 1px solid #e6e6e6;
}

.emotion-2 {
  display: none;
  background-color: #4b9fd5;
  height: 3px;
  width: 100%;
  position: absolute;
  top: -1px;
}

<BoxedTabs
  onSelect={[MockFunction]}
  selected="a"
  tabs={
    Array [
      Object {
        "key": "a",
        "label": "labela",
      },
      Object {
        "key": "b",
        "label": "labelb",
      },
    ]
  }
>
  <div>
    <Styled(button)
      active={true}
      key="0"
      onClick={[Function]}
      type="button"
    >
      <button
        className="emotion-1"
        onClick={[Function]}
        type="button"
      >
        <Styled(div)
          active={true}
        >
          <div
            className="emotion-0"
          />
        </Styled(div)>
        labela
      </button>
    </Styled(button)>
    <Styled(button)
      active={false}
      key="1"
      onClick={[Function]}
      type="button"
    >
      <button
        className="emotion-3"
        onClick={[Function]}
        type="button"
      >
        <Styled(div)
          active={false}
        >
          <div
            className="emotion-2"
          />
        </Styled(div)>
        labelb
      </button>
    </Styled(button)>
  </div>
</BoxedTabs>
`;
